test/com/google/javascript/jscomp/CollapsePropertiesTest.java

```diff
@@ -202,11 +274,22 @@ public void testGlobalObjectNameInBooleanExpressionDepth3() {
   public void testGlobalFunctionNameInBooleanExpressionDepth1() {
     test("function a() {} a.c = 1; if (a) x(a.c);",
          "function a() {} var a$c = 1; if (a) x(a$c);");
+
+    test("function a() {} /** @nocollapse */ a.c = 1; if (a) x(a.c);",
+        "function a() {} a.c = 1; if (a) x(a.c);");
   }
 
   public void testGlobalFunctionNameInBooleanExpressionDepth2() {
     test("var a = {b: function(){}}; a.b.c = 1; if (a.b) x(a.b.c);",
          "var a$b = function(){}; var a$b$c = 1; if (a$b) x(a$b$c);");
+
+    test("var a = {/** @nocollapse */ b: function(){}}; a.b.c = 1; "
+        + "if (a.b) x(a.b.c);",
+        "var a = {b: function(){}}; var a$b$c = 1; if (a.b) x(a$b$c);");
```

Comment:
If the point of nocollapse is to allow mutable properties to be exported, it's not correct to collapse the subproperties. Eg, one can mutate the uncollapsed a.b in a way that impacts a.b.c, so if we have already collapsed it to a$b$c, we produce wrong code.

Or are you expecting mutating the uncollapsed property in a restricted way, that doesn't impact subproperties? If so, do we have a well-defined rule about what we allow and what we don't?